### DISCUSSION TITLE
Importing my own segmentation model and predict in a WSI

### DISCUSSION THREAD

**Discussion by Peregalli** [AICAN-Research/FAST-Pathology#89] [Q&A] [answered]
Hi there!

I'm trying to run my own gland segmentation model with fast pathology. I convert my torch model into .onnx and now I want to see the predictions in the whole slide.  I write the following script: 
```
    importer = fast.WholeSlideImageImporter.create(wsi_path)

    tissueSegmentation = fast.TissueSegmentation.create(threshold = 70).connect(importer)

    patchGenerator = fast.PatchGenerator.create(patchSize, 
                                                patchSize, 
                                                magnification=magnification, 
                                                overlapPercent=overlapPercent)\
                                                    .connect(0, importer)\
                                                        .connect(1, tissueSegmentation)

    # Create neural network object ###
    nn = fast.NeuralNetwork.create(scaleFactor=scaleFactor,
                                   modelFilename=model_path).connect(0, patchGenerator)


    # Create patch stitcher to generate pyramidal tiff output
    stitcher = fast.PatchStitcher.create().connect(0, nn)


    # Create renderers to show both original WSI and segmentation output

    wsiRenderer = fast.ImagePyramidRenderer.create().connect(0, importer)
    segmentationRenderer = fast.SegmentationRenderer.create(opacity = 0.1,borderOpacity=1, colors={1: fast.Color.Green()}).connect(stitcher)

```
And returns me the following error: 

`ERROR [139947432331072] Standard (std) exception caught in Qt event handler SWIG director method error. In method 'execute': IndexError: index 1 is out of bounds for axis 2 with size 1`. 

I supposed that NeuralNetwork is trying to read the second axis assuming that my model output is something like [-1, 2, patchsize, patchsize], however it only has one output channel, like [-1, 1, patchsize, patchsize]. So I try to set the `outputNodes` parameter from `NeuralNetwork` function as `outputNodes = fast.NeuralNetworkNode(name = 'Conv2d-76',shape = fast.TensorShape((-1,1,patchSize,patchSize)))
` but still not working.  Does anyone have an idea of how to solve this? Thank you very much in advance!

**Comment by smistad** [comment]
I don't understand why you get a " SWIG director method error" since you are not using any custom python process objects (fast.PythonProcessObject).

However, looking at your code I see that you are using fast.NeuralNetwork, while you should instead use fast.SegmentationNetwork here.
Also, seeing that you are using torch, remember to add a softmax layer at the end if you haven't already.

Here is a tutorial on this topic: https://fast.eriksmistad.no/python-tutorial-wsi.html#autotoc_md173

**Reply by Peregalli** [reply, marked as answer]
Thanks for your quick response! I have noticed that the error was from a previous script witch uses a python process object, but that reference help me to run it, thanks!
